wgpu/src/image/raster.rs

```diff
@@ -153,9 +70,139 @@ impl Cache {
         self.get(handle).unwrap()
     }
 
-    pub fn trim(&mut self) {
+    pub fn upload(
+        &mut self,
+        handle: &image::Handle,
+        device: &wgpu::Device,
+        encoder: &mut wgpu::CommandEncoder,
+        allocator: &mut AtlasAllocator,
+        atlas: &mut wgpu::Texture,
+    ) -> &Memory {
+        let _ = self.load(handle);
+
+        let memory = self.map.get_mut(&handle.id()).unwrap();
+
+        if let Memory::Host(image) = memory {
+            let (width, height) = image.dimensions();
+            let size = Size::new(width as i32, height as i32);
+
+            let old_atlas_size = allocator.size();
+            let allocation;
+
+            loop {
+                if let Some(a) = allocator.allocate(size) {
+                    allocation = a;
+                    break;
+                }
+
+                allocator.grow(allocator.size() * 2);
+            }
+
+            let new_atlas_size = allocator.size();
+
+            if new_atlas_size != old_atlas_size {
```

Comment:
The resize logic seems to be duplicated. The new `Atlas` type will probably help here.